### PR TITLE
[SPARK-19279] [SQL] Infer Schema for Hive Serde Tables and Block Creating a Hive Table With an Empty Schema

### PR DESCRIPTION
**Pull request by gatorsmile**
### What changes were proposed in this pull request?
So far, we allow users to create a table with an empty schema: `CREATE TABLE tab1`. This could break many code paths if we enable it. Thus, we should follow Hive to block it. 

For Hive serde tables, some serde libraries require the specified schema and record it in the metastore. To get the list, we need to check `hive.serdes.using.metastore.for.schema,` which contains a list of serdes that require user-specified schema. The default values are

- org.apache.hadoop.hive.ql.io.orc.OrcSerde
- org.apache.hadoop.hive.serde2.lazy.LazySimpleSerDe
- org.apache.hadoop.hive.serde2.columnar.ColumnarSerDe
- org.apache.hadoop.hive.serde2.dynamic_type.DynamicSerDe
- org.apache.hadoop.hive.serde2.MetadataTypedColumnsetSerDe
- org.apache.hadoop.hive.serde2.columnar.LazyBinaryColumnarSerDe
- org.apache.hadoop.hive.ql.io.parquet.serde.ParquetHiveSerDe
- org.apache.hadoop.hive.serde2.lazybinary.LazyBinarySerDe

### How was this patch tested?
Added test cases for both Hive and data source tables

### CODE REVIEW

**Comment by gatorsmile** [comment]
uh, not all the serde need the schema. We need to check `hive.serdes.using.metastore.for.schema`, which contains a list of serdes that require user-specified schema:
```
        "org.apache.hadoop.hive.ql.io.orc.OrcSerde," +
        "org.apache.hadoop.hive.serde2.lazy.LazySimpleSerDe," +
        "org.apache.hadoop.hive.serde2.columnar.ColumnarSerDe," +
        "org.apache.hadoop.hive.serde2.dynamic_type.DynamicSerDe," +
        "org.apache.hadoop.hive.serde2.MetadataTypedColumnsetSerDe," +
        "org.apache.hadoop.hive.serde2.columnar.LazyBinaryColumnarSerDe," +
        "org.apache.hadoop.hive.ql.io.parquet.serde.ParquetHiveSerDe," +
        "org.apache.hadoop.hive.serde2.lazybinary.LazyBinarySerDe"
```

**Comment by gatorsmile** [comment]
retest this please

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveExternalCatalog.scala [hunk 1]
```diff
@@ -129,6 +130,21 @@ private[spark] class HiveExternalCatalog(conf: SparkConf, hadoopConf: Configurat
     }
   }
 
+  /**
+   * For Hive serde tables, some serde libraries require the specified schema and record it
+   * in the metastore. The main code is from the function hasMetastoreBasedSchema in Table.java
```
Comment (gatorsmile) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveExternalCatalog.scala]: To the reviewers: `hasMetastoreBasedSchema` in `Table.java` can be accessed via the link:  https://github.com/apache/hive/blob/master/ql/src/java/org/apache/hadoop/hive/ql/metadata/Table.java#L943-L946

**Comment by gatorsmile** [comment]
cc @cloud-fan

File: sql/core/src/test/scala/org/apache/spark/sql/execution/command/DDLSuite.scala [hunk 2]
```diff
@@ -1527,6 +1527,21 @@ class DDLSuite extends QueryTest with SharedSQLContext with BeforeAndAfterEach {
     }
   }
 
+  test("create a data source table without schema") {
+    import testImplicits._
+    withTempPath { tempDir =>
+      withTable("tab1", "tab2") {
+        (("a", "b") :: Nil).toDF().write.json(tempDir.getCanonicalPath)
+
+        val e = intercept[AnalysisException] { sql("CREATE TABLE tab1 USING json") }.getMessage
```
Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/command/DDLSuite.scala]: we should also test a data source that can infer schema without files(e.g. LibSVM data source has fixed schema). Ideally we should only fail if the given data source can't infer schema without files.

**Comment by cloud-fan** [comment]
Ideally the table schema must be specified or inferred before saving to metastore, however, for hive serde tables, we have to save it to metastore first, and let the hive metastore to infer the schema. Is it possible we can extract the schema inference logic from hive metastore? so that we can make data source tables and hive serde tables more consistent.

**Comment by gatorsmile** [comment]
: ) Done. Found a solution to infer the schema of Hive Serde tables. Let me clean the code now.

File: sql/core/src/test/scala/org/apache/spark/sql/execution/command/DDLSuite.scala (hunk 2 above)
Comment (gatorsmile) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/command/DDLSuite.scala]: This error message is not from the code added by this PR. It is from [the original logics](https://github.com/apache/spark/blob/master/sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/DataSource.scala#L343-L345). The error message is right if our file-based data sources are unable to infer the schema.

Sure, I will add a test case for `LibSVM `

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveUtils.scala [hunk 3]
```diff
@@ -455,4 +462,133 @@ private[spark] object HiveUtils extends Logging {
...
+    // maps and lists should be set only after all elements are ready (see HIVE-7975)
+    partition.storage.serde.foreach(serdeInfo.setSerializationLib)
+
+    val serdeParameters = new java.util.HashMap[String, String]()
+    catalogTable.storage.properties.foreach { case (k, v) => serdeParameters.put(k, v) }
+    partition.storage.properties.foreach { case (k, v) => serdeParameters.put(k, v) }
+    serdeInfo.setParameters(serdeParameters)
+
+    new HivePartition(hiveTable, tPartition)
+  }
```
Comment (gatorsmile) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveUtils.scala]: Basically, the above two functions `toHivePartition`, `toHiveTable` and `toHiveColumn ` are moved from `MetastoreRelation.scala`. 

`inferSchema ` and `fromHiveColumn ` are newly added.

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveUtils.scala [hunk 4]
```diff
@@ -455,4 +462,133 @@ private[spark] object HiveUtils extends Logging {
...
+  def inferSchema(table: CatalogTable): CatalogTable = {
+    if (DDLUtils.isDatasourceTable(table) || table.schema.nonEmpty) {
+      table
+    } else {
+      val hiveTable = toHiveTable(table)
+      // Note: Hive separates partition columns and the schema, but for us the
+      // partition columns are part of the schema
+      val partCols = hiveTable.getPartCols.asScala.map(fromHiveColumn)
+      val schema = StructType(hiveTable.getCols.asScala.map(fromHiveColumn) ++ partCols)
+      table.copy(schema = schema, partitionColumnNames = partCols.map(_.name))
```
Comment (gatorsmile) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveUtils.scala]: To infer the schema of Hive serde tables is simple. We just need to construct the Hive's Table struct and then call `getCols` to get the inferred schema.

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveUtils.scala [hunk 5]
```diff
@@ -455,4 +462,133 @@ private[spark] object HiveUtils extends Logging {
...
+    val metadata = new MetadataBuilder().putString(HiveUtils.hiveTypeString, hc.getType).build()
+    val field = StructField(
+      name = hc.getName,
+      dataType = columnType,
+      nullable = true,
+      metadata = metadata)
+    Option(hc.getComment).map(field.withComment).getOrElse(field)
+  }
+
+  // TODO: merge this with HiveClientImpl#toHiveTable
```
Comment (gatorsmile) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveUtils.scala]: So far, it is a little bit tricky when merging them, because our execution is using 1.2.1, but Hive metadata APIs support the versions from 0.12 to 1.2. Thus, it does not make sense to do it. 

So far, the schema inference is not using metadata Hive client. I checked the code. The changes between 0.12 and 1.2 look fine to me. Schema inference should work correctly. I think I need to add a test case to VersionSuite.scala.

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveExternalCatalog.scala [hunk 6]
```diff
@@ -26,6 +26,7 @@ import scala.util.control.NonFatal
 
 import org.apache.hadoop.conf.Configuration
 import org.apache.hadoop.fs.{FileSystem, Path}
+import org.apache.hadoop.hive.conf.HiveConf
```
Comment (cloud-fan) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveExternalCatalog.scala]: unnecessary change.

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveUtils.scala (hunk 4 above)
Comment (cloud-fan) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveUtils.scala]: will `partitionColumnNames` change?

Comment (gatorsmile) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveUtils.scala]: I think it should not be changed. Let me remove it.

File: sql/core/src/test/scala/org/apache/spark/sql/execution/command/DDLSuite.scala (hunk 2 above)
Comment (gatorsmile) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/command/DDLSuite.scala]: Test case for LibSVM is added.

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveUtils.scala (hunk 5 above)
Comment (gatorsmile) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveUtils.scala]: I moved a test case `SPARK-13709: reading partitioned Avro table with nested schema` to `VersionSuite.scala`. This can help us to know the above statement holds.

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/QueryPartitionSuite.scala [hunk 7]
```diff
@@ -68,82 +68,4 @@ class QueryPartitionSuite extends QueryTest with SQLTestUtils with TestHiveSingl
       sql("DROP TABLE IF EXISTS createAndInsertTest")
     }
   }
-
-  test("SPARK-13709: reading partitioned Avro table with nested schema") {
```
Comment (gatorsmile) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/QueryPartitionSuite.scala]: Moved it to `VersionSuite` to ensure all the Hive metastore versions can correctly infer the schema.

File: mllib/src/test/scala/org/apache/spark/ml/source/libsvm/LibSVMRelationSuite.scala [hunk 8]
```diff
@@ -110,4 +110,22 @@ class LibSVMRelationSuite extends SparkFunSuite with MLlibTestSparkContext {
...
   }
+
+  test("create libsvmTable table without schema") {
+    try {
+      spark.sql(
+        s"""
+           |CREATE TABLE libsvmTable
+           |USING libsvm
+           |OPTIONS (
+           |  path '$path'
```
Comment (cloud-fan) [inline mllib/src/test/scala/org/apache/spark/ml/source/libsvm/LibSVMRelationSuite.scala]: what if no path specified?

Comment (gatorsmile) [inline mllib/src/test/scala/org/apache/spark/ml/source/libsvm/LibSVMRelationSuite.scala]: It will output an error message `No input path specified for libsvm data`. Added a test case for it.

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/HiveDDLSuite.scala [hunk 9]
```diff
@@ -79,6 +79,25 @@ class HiveDDLSuite
...
+    withTempPath { tempDir =>
+      withTable("tab1", "tab2") {
+        (("a", "b") :: Nil).toDF().write.json(tempDir.getCanonicalPath)
+
+        var e = intercept[AnalysisException] { sql("CREATE TABLE tab1 USING hive") }.getMessage
+        assert(e.contains("Unable to infer the schema. The schema specification is required to " +
+          "create the table `default`.`tab1`"))
+
+        e = intercept[AnalysisException] {
+          sql(s"CREATE TABLE tab2 location '${tempDir.getCanonicalPath}'")
```
Comment (cloud-fan) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/HiveDDLSuite.scala]: is there any hive serde that can infer schema from data files?

**Comment by cloud-fan** [comment]
LGTM, merging to master!

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/HiveDDLSuite.scala (hunk 9 above)
Comment (gatorsmile) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/HiveDDLSuite.scala]: Based on the official Hive document about [Row Formats & SerDe](https://cwiki.apache.org/confluence/display/Hive/LanguageManual+DDL#LanguageManualDDL-RowFormats&SerDe), users must specify the columns of the schema for the native SerDe. For the custom SerDe, Hive will query the SerDe to get the columns in the schema. It does not sounds Hive has such a capability to infer the schema from the data files. 

Actually, there is two unresolved tickets for it. https://issues.apache.org/jira/browse/HIVE-8950 and https://issues.apache.org/jira/browse/HIVE-10593. It sounds like the Hive community prefers to adding a new DDL syntax to get the table schema from a Parquet file (or some other format), but it is not available in Hive yet

Comment (gatorsmile) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/HiveDDLSuite.scala]: Found an interesting case. I am thinking we need to support what Hive did in `CREATE TABLE LIKE`. That is, allowing users to create another table with a different data source provider:
```
CREATE TABLE avro_test ROW FORMAT SERDE 'org.apache.hadoop.hive.serde2.avro.AvroSerDe' STORED AS AVRO TBLPROPERTIES ('avro.schema.url'='myHost/myAvroSchema.avsc'); 
CREATE EXTERNAL TABLE parquet_test LIKE avro_test STORED AS PARQUET LOCATION 'hdfs://myParquetFilesPath';
```
